lib/websocket/connection.js

```diff
@@ -0,0 +1,321 @@
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+
+  if (channels.open.hasSubscribers) {
+    channels.open.publish({
+      address: response.socket.address(),
+      protocol,
+      extensions
+    })
+  }
+}
+
+/**
+ * @param {Buffer} chunk
+ */
+function onSocketData (chunk) {
+  if (!this.ws[kByteParser].write(chunk)) {
+    this.pause()
+  }
+}
+
+function onParserDrain () {
+  this.ws[kResponse].socket.resume()
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @see https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.4
+ */
+function onSocketClose () {
+  const { ws } = this
+
+  // If the TCP connection was closed after the
+  // WebSocket closing handshake was completed, the WebSocket connection
+  // is said to have been closed _cleanly_.
+  const wasClean = ws[kSentClose] && ws[kReceivedClose]
+
+  let code = 1005
+  let reason = ''
+
+  const result = ws[kByteParser].closingInfo
+
+  if (result) {
+    code = result.code ?? 1005
+    reason = result.reason
+  } else if (!ws[kSentClose]) {
+    // If _The WebSocket
+    // Connection is Closed_ and no Close control frame was received by the
+    // endpoint (such as could occur if the underlying transport connection
+    // is lost), _The WebSocket Connection Close Code_ is considered to be
+    // 1006.
+    code = 1006
+  }
+
+  // 1. Change the ready state to CLOSED (3).
+  ws[kReadyState] = states.CLOSED
+
+  // 2. If the user agent was required to fail the WebSocket
+  //    connection, or if the WebSocket connection was closed
+  //    after being flagged as full, fire an event named error
+  //    at the WebSocket object.
+  // TODO
+
+  // 3. Fire an event named close at the WebSocket object,
+  //    using CloseEvent, with the wasClean attribute
+  //    initialized to true if the connection closed cleanly
+  //    and false otherwise, the code attribute initialized to
+  //    the WebSocket connection close code, and the reason
+  //    attribute initialized to the result of applying UTF-8
+  //    decode without BOM to the WebSocket connection close
+  //    reason.
+  fireEvent('close', ws, CloseEvent, {
+    wasClean, code, reason
+  })
+
+  if (channels.close.hasSubscribers) {
+    channels.close.publish({
+      websocket: ws,
+      code,
+      reason
+    })
+  }
+}
+
+function onSocketError (error) {
+  const { ws } = this
+
+  fireEvent('error', ws, ErrorEvent, {
```

Comment:
This might cause multiple `'error'` events because there is another place where this is called. It does not cause issues in practice because `ws` is an `EventTarget` but I'm not sure if it is wanted. To match the behavior of browsers and the `http` module, in `websockets/ws` socket errors are not returned to the user.

Reply:
Added diagnostic channel support instead.